Userland/Libraries/LibWeb/XHR/FormData.cpp

```diff
@@ -0,0 +1,170 @@
+ */
+
+#include <AK/TypeCasts.h>
+#include <LibJS/Runtime/Completion.h>
+#include <LibWeb/Bindings/Intrinsics.h>
+#include <LibWeb/FileAPI/Blob.h>
+#include <LibWeb/FileAPI/File.h>
+#include <LibWeb/HTML/FormControlInfrastructure.h>
+#include <LibWeb/WebIDL/DOMException.h>
+#include <LibWeb/XHR/FormData.h>
+
+namespace Web::XHR {
+
+// https://xhr.spec.whatwg.org/#dom-formdata
+WebIDL::ExceptionOr<JS::NonnullGCPtr<FormData>> FormData::construct_impl(JS::Realm& realm, Optional<JS::NonnullGCPtr<HTML::HTMLFormElement>>)
+{
+    // FIXME: 1. If form is given, then:
+    // FIXME:    1. Let list be the result of constructing the entry list for form.
+    // FIXME:    2. If list is null, then throw an "InvalidStateError" DOMException.
+    // FIXME:    3. Set this’s entry list to list.
+
+    return construct_impl(realm);
+}
+
+WebIDL::ExceptionOr<JS::NonnullGCPtr<FormData>> FormData::construct_impl(JS::Realm& realm, HashMap<DeprecatedString, Vector<Variant<JS::NonnullGCPtr<FileAPI::File>, DeprecatedString>>> entry_list)
+{
+    return MUST_OR_THROW_OOM(realm.heap().allocate<FormData>(realm, realm, move(entry_list)));
+}
+
+FormData::FormData(JS::Realm& realm, HashMap<DeprecatedString, Vector<Variant<JS::NonnullGCPtr<FileAPI::File>, DeprecatedString>>> entry_list)
+    : PlatformObject(realm)
+    , m_entry_list(move(entry_list))
+{
+    set_prototype(&Bindings::ensure_web_prototype<Bindings::FormDataPrototype>(realm, "FormData"));
+}
+
+FormData::~FormData() = default;
+
+void FormData::visit_edges(Cell::Visitor& visitor)
+{
+    Base::visit_edges(visitor);
+    for (auto const& entry : m_entry_list) {
+        for (auto const& value : entry.value) {
+            if (auto* file = value.get_pointer<JS::NonnullGCPtr<FileAPI::File>>())
+                visitor.visit(*file);
+        }
+    }
+}
+
+// https://xhr.spec.whatwg.org/#dom-formdata-append
+WebIDL::ExceptionOr<void> FormData::append(DeprecatedString const& name, DeprecatedString const& value)
+{
+    auto& vm = realm().vm();
+    return inner_append(TRY_OR_THROW_OOM(vm, String::from_deprecated_string(name)), TRY_OR_THROW_OOM(vm, String::from_deprecated_string(value)));
+}
+
+// https://xhr.spec.whatwg.org/#dom-formdata-append-blob
+WebIDL::ExceptionOr<void> FormData::append(DeprecatedString const& name, JS::NonnullGCPtr<FileAPI::Blob> const& blob_value, Optional<DeprecatedString> const& filename)
+{
+    auto& vm = realm().vm();
+    auto inner_filename = filename.has_value() ? TRY_OR_THROW_OOM(vm, String::from_deprecated_string(filename.value())) : Optional<String> {};
+    return inner_append(TRY_OR_THROW_OOM(vm, String::from_deprecated_string(name)), blob_value, inner_filename);
+}
+
+// https://xhr.spec.whatwg.org/#dom-formdata-append
+// https://xhr.spec.whatwg.org/#dom-formdata-append-blob
+WebIDL::ExceptionOr<void> FormData::inner_append(String const& name, Variant<JS::NonnullGCPtr<FileAPI::Blob>, String> const& value, Optional<String> const& filename)
```

Comment:
let's call this `append_impl`, seems closer to our usual naming for delegated implementations :)